numpy/core/src/multiarray/scalarapi.c

```diff
@@ -390,42 +391,51 @@ PyArray_ScalarFromObject(PyObject *object)
             PyArrayScalar_RETURN_FALSE;
         }
     }
-    else if (PyInt_Check(object)) {
-        ret = PyArrayScalar_New(Long);
-        if (ret == NULL) {
-            return NULL;
+    else if (PyLong_Check(object)) {
+        npy_longlong val;
+
+        /* Check if fits in long */
+        val = (npy_longlong)PyLong_AsLong(object);
+        if (!error_converting(val)) {
+            ret = PyArrayScalar_New(Long);
+            if (ret != NULL) {
+                PyArrayScalar_VAL(ret, Long) = (npy_long)val;
+            }
+            return ret;
+        }
+        PyErr_Clear();
+
+        /* Check if fits in long long */
+        val = PyLong_AsLongLong(object);
+        if (!error_converting(val)) {
+            ret = PyArrayScalar_New(LongLong);
+            if (ret != NULL) {
+                PyArrayScalar_VAL(ret, LongLong) = val;
+            }
+            return ret;
         }
-        PyArrayScalar_VAL(ret, Long) = PyLong_AsLong(object);
+        PyErr_Clear();
```

Comment:
Doesn't clearing this result in a SystemError?

Reply:
How so? `PyErr_Clear()` has no effect if there is no error.

Reply:
I was motivated by the comment up top
```
/*NUMPY_API
 * Get an Array Scalar From a Python Object
 *
 * Returns NULL if unsuccessful but error is only set if another error occurred.
 * Currently only Numeric-like object supported.
 */
NPY_NO_EXPORT PyObject *
PyArray_ScalarFromObject(PyObject *object)
```

Reply:
Take a look at the full original, it also cleared the error for the `long long` case.

Reply:
Probably right to leave it as is, NULL returns without error tend to be a bit strange, but here its probably correct: It indicates fallback to a default path.

The confusing part may be that this function is really an internal helper for `scalarmath.c.src` and I expect only ever exposed due to the umath/multiarray split.

Reply:
> I was motivated by the comment up top

I didn't spot that, thanks for confirming this is fine.

Reply:
Note that `PyErr_NoMemory` is not cleared.

Reply:
> Note that `PyErr_NoMemory` is not cleared.

What do you mean by that?

Reply:
There may be allocation errors for the scalar instances. I didn't check NumPy specifically, but some of the `tp_alloc` functions in Python can fail that way and set an error.